ompi/mca/mtl/psm2/mtl_psm2_component.c

```diff
@@ -89,6 +93,30 @@ ompi_mtl_psm2_component_register(void)
 
     /* set priority high enough to beat ob1's default (also set higher than psm) */
     param_priority = 40;
+#if OPAL_CUDA_SUPPORT
+    /*
+     * If using CUDA enabled OpenMPI, the user likely intends to
+     * run with CUDA buffers. So, force-set the envvar here if user failed
+     * to set it.
+     */
+    cuda_env = getenv("PSM2_CUDA");
+    if (!cuda_env) {
+        opal_show_help("help-mtl-psm2.txt",
+                       "no psm2 cuda env", true,
+                       "not set",
+                       "Host buffers,\nthere will be a performance penalty"
+                       " due to OMPI force setting this variable now.\n"
+                       "Set environment variable to 0 if using Host buffers" );
```

Comment:
Do not use "OMPI" in a show help string -- the product name is "Open MPI" (two words).

I would suggest that you should have 2 different show_help messages here, rather than feeding in a lengthy string to `opal_show_help()`.  

Additionally -- this message is being displayed during the component registration phase.  Isn't that far too early?  I.e., won't it display on systems that don't even have PSM2 and/or CUDA hardware available?

I would think that you should only display these messages if a) the PSM2 MTL is actually selected *and* b) there are CUDA devices in the system.  Otherwise, if this message is emitted if a) the user doesn't have PSM2 hardware or b) doesn't have CUDA devices, the user will (rightfully) be quite confused.  For example, you might replace this message with:

> Warning: Open MPI has detected both PSM2-based hardware and CUDA hardware, but the PSM2_CUDA environment variable was not set.  Open MPI has therefore defaulted to setting PSM2_CUDA=1.
>
> This can leave to a performance penalty, however, if your application uses host buffers.  You should set the PSM2_CUDA environment variable to 0 before invoking mpirun to both silence this warning and provide the hint to Open MPI / PSM2 that your application will mainly be using host buffers.
>
>   Local hostname: %s

(and include the local hostname in the message, just to let the user know which machine Open MPI is talking about)

You can have a similar message for the 2nd `opal_show_help()`, below.

Also, keep in mind that the component registration function is really only intended to be used to register MCA vars -- it is not actually intended to *do* anything in terms of initialization or setup.

If you *really really really* have to set the `PSM2_CUDA` environment variable all the way up here in the component registration function, you should include a lengthy comment about *why* you have to do so.  And then you'll need to set some kind of state variable (on the component?) indicating that you do so so that you can emit an appropriate show_help message later (i.e., if/when the PSM2 MTL is selected to be used).

Make sense?

Reply:
@jsquyres , will fix the help text output in separate commit against master and then cherry pick here.